### PR TITLE
Refactor presenter preload

### PR DESCRIPTION
**Pull request by robyurkowski**
I need to add one more fix here before this is merged (missing a test for when there are no valid preloads), but I think this is a reasonable attempt at removing some of the complexity of the presenter into an appropriate collaborator.

### CODE REVIEW

File: lib/brainstem/preloader.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+    def initialize(models, preloads, reflections, preload_method = nil)
+      self.models         = models
+      self.preloads       = preloads.compact
+      self.reflections    = reflections
+      self.preload_method = preload_method
+      self.valid_preloads = {}
+    end
+
```
Comment (cantino) [inline lib/brainstem/preloader.rb]: This is more whitespace between definitions than we typically use.

File: lib/brainstem/presenter.rb [hunk 2]
```diff
@@ -53,11 +54,6 @@ def self.reflections(klass)
     end
 
     def self.ar_preload(models, association_names)
-      if Gem.loaded_specs['activerecord'].version >= Gem::Version.create('4.1')
```
Comment (cantino) [inline lib/brainstem/presenter.rb]: why was this removed?  how is it passing without it?

File: lib/brainstem/preloader.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    def preload!
+      preload_method.call(models, valid_preloads) if valid_preloads.keys.any?
+    end
+
+
+    # Returns a proc that takes two arguments, +models+ and +association_names+,
+    # which when called preloads those.
```
Comment (cantino) [inline lib/brainstem/preloader.rb]: "which, when called, preloads those."

File: spec/brainstem/preloader_spec.rb [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    context "when preload in the list of reflections" do
+      let(:reflections) { { users: [], posts: [] } }
+
+      it "keeps it" do
+        expect(preloader.valid_preloads.keys).to eq [:users, :posts]
+      end
+    end
+
+    context "when preload not in the list of reflections" do
+      let(:reflections) { { users: [] } }
```
Comment (cantino) [inline spec/brainstem/preloader_spec.rb]: pretty sure reflections has string keys, not symbols.

File: spec/brainstem/presenter_spec.rb [hunk 5]
```diff
@@ -404,41 +404,73 @@ def workspace.synthetic_id
...
+      #    inappropriately here -- a fact that I think bears weight given
+      #    we're making unit-level assertions in an integration spec.
+      #    However, there's also some value in asserting that these are
+      #    passed through without digging into Rails internals. However,
+      #    further 'purity' improvements could be made by introspecting on
+      #    AR's actual data structures.
+      #
+      #    That's about three shades on the side of overkill, though.
+      #
+      def preloader_should_receive(hsh)
```
Comment (cantino) [inline spec/brainstem/presenter_spec.rb]: Is all of this to get around stubbing the old `ar_preload`?

File: lib/brainstem/presenter.rb (hunk 2 above)
Comment (robyurkowski) [inline lib/brainstem/presenter.rb]: Looks like I forgot to yank the surrounding method somehow. This is injected in Preloader as a proc.

File: lib/brainstem/preloader.rb (hunk 1 above)
Comment (robyurkowski) [inline lib/brainstem/preloader.rb]: Tightened that up.

File: spec/brainstem/presenter_spec.rb (hunk 5 above)
Comment (robyurkowski) [inline spec/brainstem/presenter_spec.rb]: I've moved `ar_preload` into the Preloader as the presenter shouldn't care about the actual method of preloading. 

Inside `Preloader`, it's appropriate to inject a method or a proc as a collaborator because it should be insulated from the actual implementation within ActiveRecord—or, put another way, what the preloader needs is to send a message to ActiveRecord, but just *which* message is actually an implementation detail of AR and should be injected as a dependency. 

We could move the `ar_preload` method entirely within the Preloader, but that closely binds Preloader to AR. In some situations (testing being the biggest example), we want to inject a different collaborator that handles the way that preloading is done.

**Comment by robyurkowski** [comment]
@cantino I've addressed your feedback here -- if this is good to go, I'll squash before merge.

**Comment by cantino** [comment]
Go ahead and squash.

**Comment by robyurkowski** [comment]
Squashed / rebased.

**Comment by cantino** [comment]
Thanks!
